### DISCUSSION TITLE
Returning false won't stop a record being created with onRecordBeforeCreateRequest()

### DISCUSSION THREAD

**Discussion by lia-07** [pocketbase/pocketbase#3610] [Q&A] [answered]
Hi all, 
I am working on a non-critical chat app with PocketBase and want to add a server-side rate limit.
The chat app doesn't have any authentication (by choice) so I am using browser fingerprints in their place (I know this is insecure and can be spoofed, and I don't care – as I said, this is non-critical and more of an experiment). 
Because the VPS I am using is slow (it became unresponsive for 10 minutes when I tried building PocketBase) I decided to use JavaScript to extend PocketBase to implement a rate limit. 

My script works by checking if the last record in the messages collection where the browser fingerprint is the same as the request's was created more than 3000 milliseconds ago. It's probably not the most elegant solution, but it's all I could think of.

I had to do some very peculiar work arounds to get the desired behaviour out of Goja (mainly `const record = JSON.parse(JSON.stringify(e.record));`), but so far my script seems to be working (if you send more than one message in three seconds, it does print "RATE LIMIT"). Despite this, the `return false` beneath that console.log() doesn't seem to do anything. The record still gets created, and the client still gets sent a success response. The same problem happens when I throw an error there as well.

Am I missing something? Any help would be appreciated.

Here is my code:
```javascript
// pb_hooks/main.pb.js

onRecordBeforeCreateRequest(async (e) => {
  // don't even ask...
  const record = JSON.parse(JSON.stringify(e.record));

  const browserFingerprint = record.browser_fingerprint;

  console.log(browserFingerprint);
  async function getLastMessage(browserFingerprint) {
    try {
      const lastMessage = await $app
        .dao()
        .findRecordsByFilter(
          "messages",
          "browser_fingerprint = {:fingerprint}",
          "-created",
          1,
          0,
          {
            fingerprint: browserFingerprint,
          }
        );
      return lastMessage[0];
    } catch (error) {
      console.error("An error occurred: ", error);
    }
  }

  const lastMessage = await getLastMessage(browserFingerprint);

  if (lastMessage) {
    console.log(
      "Previous message from browser: " + JSON.stringify(lastMessage)
    );

    // what is goja on bro
    var lastMessageTime = new Date(
      lastMessage.created.toString().replace(" ", "T")
    );

    var now = new Date();

    var msSinceLastMessage = now - lastMessageTime;

    if (msSinceLastMessage < 3000) {
      console.log("RATE LIMIT");
      return false;
    }
  }
}, "messages");
```

Thanks for reading!

**Comment by ganigeorgiev** [comment, marked as answer]
You are returning a `Promise`, not `false`.

Remove the async/await calls as it doesn't really have any use with the PocketBase APIs since they are synchronous (_the hook handler is non-blocking as it is executed in their own goroutine_).

Also if you want a map you can replace `JSON.parse/JSON.stringify` with just `e.record.publicExport()`.
If you just want to access a record field value you can call directly `e.record.get("browser_fingerprint")`.

**Reply by lia-07** [reply]
Thanks for the quick reply.
I'm not sure I understand how I am returning a `Promise`, and not a `false` value. How would I fix this?

**Reply by ganigeorgiev** [reply]
Prefixing the hook handler function with `async` will make it to return a `Promise`.
To fix it remove all async/await statements from the above code.

**Reply by lia-07** [reply]
Thank you so much for your help. My apologies for the obvious questions!
